### PR TITLE
Jss authentication

### LINKED ISSUE #12: Authentication
#### As an application hoster I want to add user authentication to my application.

__Story__:

We have done a great job providing immediate value to our customers. At this
point they can both add and delete images, in addition to a few other features
that are either completed or in the works. Along with a growth in our web
service's popularity have come an increase in number of reports of missing
images. Upon investigation, it turns out that the majority of delete requests
are issued by only a handful of IP addresses. While we could blacklist those IP
addresses from the delete operation, it has always been our plan to only permit
the user who added an image to delete that image. To begin that effort we must
first add the concept of a user.

Please split this task up into three separate PRs. The second, and third should
depend on the preceding ones.

__Acceptance criteria PR1__:
- [x] Users can sign up.

__Acceptance criteria PR2__:
- [x] Users can log in after they've signed up.
- [x] Users can log out after logging in.
- [ ] Logged in users should not be able to register, nor log in again.

__Acceptance criteria PR3__:
- [ ] Users can choose to be remembered as part of the login
  process.<sup>†</sup>
- [ ] Multiple browsers can concurrently maintain the remembered status for the
  same user.

<sup>†</sup>: "Remembering" means that users will remain logged into the site
even after completely closing their web browser. Please be sure to completely
close the browser (`cmd` + `q`) when testing this functionality. Closing the
window (`cmd` + `w`, or clicking the red circle) on Mac OS does not close the
browser.

__Note__: Do not use Devise for this task.

__Discussion topics__:
- [ ] If we wanted to add more advanced features such as email activation and
  "forgot my password", should we continue to "roll our own" authentication?

__Reference material__:
- https://www.railstutorial.org/book/modeling_users
- https://www.railstutorial.org/book/sign_up
- https://www.railstutorial.org/book/basic_login
- https://www.railstutorial.org/book/advanced_login

__Further reading__:
- https://github.com/plataformatec/devise

__Dependencies__:
- Delete Image
- Refactor flow tests to use page objects

### PR DESCRIPTION
**Pull request by jschneid**
Closes #12

### CODE REVIEW

File: db/migrate/20171108182257_create_users.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class CreateUsers < ActiveRecord::Migration[5.0]
+  def change
+    create_table :users do |t|
+      t.string :name, null: false
+      t.string :email, null: false
+      t.string :password_digest
```
Comment (bboe) [inline db/migrate/20171108182257_create_users.rb]: In what cases can the password_digest be `null`?

File: test/fixtures/users.yml [hunk 2]
```diff
@@ -0,0 +1 @@
+# Read about fixtures at http://api.rubyonrails.org/classes/ActiveRecord/FixtureSet.html
```
Comment (bboe) [inline test/fixtures/users.yml]: Commenting here because I cannot comment on a blank file:

image

Should the empty file be part of this commit?

File: test/models/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  def test_user_new
+    user = User.new(name: 'jon', email: 'jon@example.com', password: 'password1', password_confirmation: 'password1')
```
Comment (bboe) [inline test/models/user_test.rb]: Can you DRY up your calls to `User.new` such that you have a valid object by default, but can overwrite individual fields in order to make those invalid?

File: test/models/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    assert_predicate user, :valid?
+  end
+
+  def test_email_coerced_to_lowercase
+    user = User.new(name: 'jon', email: 'JON@Example.com', password: 'password1', password_confirmation: 'password1')
+    user.save
```
Comment (bboe) [inline test/models/user_test.rb]: While it's rare, consider using `save!` so that an exception raises in the event this happens to fail. Similar for `create!` and `update!`. The non-`!` should be used only in cases where you are explicitly utilizing the return value.

File: test/models/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
+require 'test_helper'
+
+class UserTest < ActiveSupport::TestCase
+  def test_user_new
+    user = User.new(name: 'jon', email: 'jon@example.com', password: 'password1', password_confirmation: 'password1')
+    assert_predicate user, :valid?
+  end
+
+  def test_email_coerced_to_lowercase
```
Comment (bboe) [inline test/models/user_test.rb]: Great test case.

File: test/models/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    user = User.new(name: 'jon', email: 'JON@Example.com', password: 'password1', password_confirmation: 'password1')
+    user.save
+    assert_equal 'jon@example.com', user.email
+  end
+
+  def test_user_name_required
+    user = User.new(name: '', email: 'jon@example.com', password: 'password1', password_confirmation: 'password1')
+    assert_predicate user, :invalid?
```
Comment (bboe) [inline test/models/user_test.rb]: When a model is not valid error messages are attached to the model. For all of your invalid test cases, can you assert that only the exact error message you expect is set?

File: test/models/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  def test_user_email_required
+    user = User.new(name: 'jon', email: '', password: 'password1', password_confirmation: 'password1')
+    assert_predicate user, :invalid?
+  end
+
+  def test_user_name_max_length
```
Comment (bboe) [inline test/models/user_test.rb]: Great test case! Can we test the successful boundary condition as well? Doing so ensures that the length cannot be trivially changed without requiring updating a test case. Sometimes I put those in a single test function so that way it's clear it's testing the boundary. Can you also do the same for email and password?

File: test/models/user_test.rb [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  def test_email_unique
+    user1 = User.new(name: 'Jon Schneider', email: 'jon@example.com', password: 'password1', password_confirmation: 'password1')
+    user1.save
+    user2 = User.new(name: 'Jonathan Schneider', email: 'jon@example.com', password: 'password2', password_confirmation: 'password2')
+    assert_predicate user2, :invalid?
+  end
+
+  def test_email_unique_case_insensitive
```
Comment (bboe) [inline test/models/user_test.rb]: Great test case.

File: test/models/user_test.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    user2 = User.new(name: 'Jonathan Schneider', email: 'JON@example.com', password: 'password2', password_confirmation: 'password2')
+    assert_predicate user2, :invalid?
+  end
+
+  def test_user_email_well_formed
+    valid_addresses = %w[user@example.com USER@foo.COM A_US-ER@foo.bar.org
+                         first.last@foo.jp alice+bob@baz.cn]
+    valid_addresses.each do |valid_address|
+      user = User.new(name: 'jon', email: valid_address, password: 'password1', password_confirmation: 'password1')
+      assert user.valid?, "#{valid_address.inspect} should be valid"
```
Comment (bboe) [inline test/models/user_test.rb]: While I am really happy to see the second argument to `assert` so that it's not a "naked assert", I'm curious why you didn't use `assert_predicate` here, when it was being well used elsewhere.

File: test/models/user_test.rb [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      assert_not user.valid?, "#{invalid_address.inspect} should be invalid"
+    end
+  end
+
+  def test_user_password_required
+    user = User.new(name: 'jon', email: 'jon@example.com', password: '', password_confirmation: '')
+    assert_predicate user, :invalid?
+  end
+
+  def test_user_password_confirmation_mismatch
```
Comment (bboe) [inline test/models/user_test.rb]: Great test.

File: app/assets/javascripts/users.js [hunk 11]
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
Comment (bboe) [inline app/assets/javascripts/users.js]: Do we need this file?

File: app/assets/stylesheets/users.scss [hunk 12]
```diff
@@ -0,0 +1,3 @@
+// Place all the styles related to the Users controller here.
```
Comment (bboe) [inline app/assets/stylesheets/users.scss]: Do we need this file?

File: app/helpers/users_helper.rb [hunk 13]
```diff
@@ -0,0 +1,2 @@
+module UsersHelper
```
Comment (bboe) [inline app/helpers/users_helper.rb]: Do we need this file?

File: app/views/users/create.html.erb [hunk 14]
```diff
@@ -0,0 +1,2 @@
+<h1>Users#create</h1>
```
Comment (bboe) [inline app/views/users/create.html.erb]: Do we need this file?

File: test/controllers/users_controller_test.rb [hunk 15]
```diff
@@ -0,0 +1,31 @@
+require 'test_helper'
+
+class UsersControllerTest < ActionDispatch::IntegrationTest
+  def test_new__success
+    get new_user_path
+    assert_response :ok
+    assert_select '#new_user'
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: What's the important element that should be present on this page? Hint, it's the element that will ensure we end up being able to submit the data to the right location. Can we make an assertion about that element and path associated with it?

File: test/controllers/users_controller_test.rb [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  end
+
+  def test_create__success
+    assert_difference('User.count', 1) do
+      user_params = { name: 'wei-ping', email: 'wei-ping@example.com', password: 'abcd1234', password_confirmation: 'abcd1234' }
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Can we DRY up `user_params` such that it doesn't need to be explicitly written out in each test case that requires it?

File: test/controllers/users_controller_test.rb [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    assert_equal 'Your new user account was successfully created! Welcome, wei-ping!', flash[:success]
+  end
+
+  def test_create__fail
+    assert_no_difference('User.count') do
+      user_params = { name: 'wei-ping', email: 'invalid_email', password: 'abcd1234', password_confirmation: 'abcd1234' }
+      post users_path, params: { user: user_params }
+    end
+
+    assert_response :ok
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Should this be a "200 OK" response?

File: test/controllers/users_controller_test.rb [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+    assert_redirected_to root_path
+    assert_equal 'Your new user account was successfully created! Welcome, wei-ping!', flash[:success]
+  end
+
+  def test_create__fail
+    assert_no_difference('User.count') do
+      user_params = { name: 'wei-ping', email: 'invalid_email', password: 'abcd1234', password_confirmation: 'abcd1234' }
+      post users_path, params: { user: user_params }
+    end
+
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: You did a great job ensuring the User count did not increase (no database side effect). Should we also ensure there is no flash message side effect? We would expect that side effect to occur if the action were successful.

File: test/flows/users_crud_test.rb [hunk 19]
```diff
@@ -0,0 +1,38 @@
...
+    # Fill out and submit the registration form with invalid inputs; verify we ended up back on the Sign Up page
+    sign_up_user_page = sign_up_user_page.fill_and_submit_form!(
+      name: '',
+      email: 'not_a_valid_email',
+      password: 'abc',
+      password_confirmation: 'abcdefg'
+    ).as_a(PageObjects::Users::SignUpUserPage)
+
+    # Verify error messages
+    assert sign_up_user_page.name.error_message.present?
```
Comment (bboe) [inline test/flows/users_crud_test.rb]: Can you use `assert_predicate` here?

File: db/migrate/20171108182257_create_users.rb (hunk 1 above)
Comment (jschneid) [reply db/migrate/20171108182257_create_users.rb]: In the case when the developer didn't initially implement the population of the password field with a value, and then neglected to come back and change it afterward...? 😅 Updated now to enforce non-null!

**Comment by jschneid** [comment]
Regarding the email validation in the user model: To be transparent, I just copied that email validation regex straight from railtutorial.org (at https://www.railstutorial.org/book/modeling_users#table-valid_email_regex), which claims it is a "regex that has proven to be robust in practice". I figured that was good enough endorsement for the purposes of this simple app.

I'm absolutely happy, though, to both simplify the code and improve the quality of the email validation by using a trusted library/gem to do the validation!  Thanks for letting me know about it!  Added it in 7c0baa2c2e1ffeccc019ea1e96bd2c5036de4118.  

(A test to verify that that gem's email validation is present hasn't yet been added as of that changeset; I'm adding that now.)

File: test/models/user_test.rb (hunk 9 above)
Comment (jschneid) [reply test/models/user_test.rb]: This is another spot where I just copy+pasted from railstutorial.org and called it a day, since the sample code there seemed to do exactly what I needed: https://www.railstutorial.org/book/modeling_users#code-email_format_validation_tests

I agree though that it would better to be consistent in how we're doing asserts; thanks for the catch!

Fixed in 6a96db941072a4b5c92b4458166f19f25acf0839.

File: test/flows/users_crud_test.rb (hunk 19 above)
Comment (jschneid) [reply test/flows/users_crud_test.rb]: Good call, I need to form that habit! Done in 9169857ae2f6e07d267f231d30383f165f122d64.

File: test/controllers/users_controller_test.rb (hunk 15 above)
Comment (jschneid) [reply test/controllers/users_controller_test.rb]: Well, although the user clicks on a Submit button to submit the data, it's the form element itself whose attributes define what happen on submit.  It does make sense that we'd want a test failure to happen if that were to change or break.

After a bit of searching around, I couldn't come up with a good example online to mimic for this, so I took a stab myself:

```
  # Verify that the single form on the page, when submitted, does a POST to the expected route
  def test_new__form_submit
    get new_user_path
    assert_select 'form', count: 1 do
      assert_select '[method=?]', 'post'
      assert_select '[action=?]', users_path
    end
  end
```

(That's in 0240239a4b9f8a7c426f6e56fa4e885487321e7a.) How could I improve that?

Also, I wanted to do a case-insensitive compare on the method (since it's fine if it turns out to be `POST` instead of `post`, for example), but some quick searching didn't come up with a straightforward way to achieve this.  Any suggestions?  Or is rails pretty much guaranteed to always output the method in lowercase in the generated HTML, and so there's no need to worry about it?

File: app/assets/javascripts/users.js (hunk 11 above)
Comment (jschneid) [reply app/assets/javascripts/users.js]: Removed! Good suggestion, thanks!

File: app/assets/stylesheets/users.scss (hunk 12 above)
Comment (jschneid) [reply app/assets/stylesheets/users.scss]: Likewise removed!

File: app/helpers/users_helper.rb (hunk 13 above)
Comment (jschneid) [reply app/helpers/users_helper.rb]: Likewise removed!

File: app/views/users/create.html.erb (hunk 14 above)
Comment (jschneid) [reply app/views/users/create.html.erb]: Likewise removed!

File: test/controllers/users_controller_test.rb (hunk 16 above)
Comment (jschneid) [reply test/controllers/users_controller_test.rb]: Yes! Replaced those with a new `@user_params` instance variable (is that what it's called in Ruby?).

File: test/controllers/users_controller_test.rb (hunk 18 above)
Comment (jschneid) [reply test/controllers/users_controller_test.rb]: Added: `assert_predicate flash, :empty?`

File: test/controllers/users_controller_test.rb (hunk 17 above)
Comment (jschneid) [reply test/controllers/users_controller_test.rb]: I did notice, and it did strike me as a bit odd, that form errors were generating HTTP 200 responses.  At the time, I just shrugged it off as "I guess that's how rails does things!"

With you questioning this assumption, I did a bit more research, and discovered the optional `status:` parameter on `render`. Aha! I've changed it to render, and assert, a status of `:bad_request` (from http://guides.rubyonrails.org/layouts_and_rendering.html#the-status-option).

I made the same change as well on the existing "add image" form, for consistency.

File: test/fixtures/users.yml (hunk 2 above)
Comment (jschneid) [reply test/fixtures/users.yml]: It should not have been! It must have ended up like that due to me messing up an interactive rebase merge at some point.

I've tried to rewrite the project history to remove that file from that original commit (but leave it in subsequent commits), but unfortunately, my git-fu does not seem to be strong enough yet; I've tried (roughly) following instructions I found at https://stackoverflow.com/a/28173964/12484 but without success at this point.

Since this is pretty minor (and this isn't a real production project), I'm going to leave this as-is at this point. (If and when this does come up in a real project, I'll invest the additional time and/or borrow others' time to get some help to figure out how to correct the problem at that point.)

File: test/models/user_test.rb (hunk 3 above)
Comment (jschneid) [reply test/models/user_test.rb]: This is a great suggestion. In addition to being more succinct, it makes the code in additional test methods a lot more readable. Thanks!

File: test/models/user_test.rb (hunk 4 above)
Comment (jschneid) [reply test/models/user_test.rb]: Totally makes sense to fail fast in such a situation. Thanks for the suggestion! Implemented in 7be79965b5ae9fb510e320c8bdd4e83efa6a9b3d.

File: test/models/user_test.rb (hunk 6 above)
Comment (jschneid) [reply test/models/user_test.rb]: I added the relevant asserts in 7ad7e486cef09ec70c8879769c93562daf9570a5.

One question I've got at this point is that all I did was to go ahead and assert that the not-incredibly-user-friendly error messages provided by Rails (e.g. "is invalid") are what we get. In a real app, I think I'd want to take the time and enhance the model to return nicer error messages (assuming that's possible), and assert those.  Is it worth doing that for this learning app?

**Comment by jschneid** [comment]
Question: Along the lines of how we replaced trying to use a custom (albeit copied from the docs) email validation regex with a gem designed for that purpose, is there a good, standard _url_ validation gem out there? I've currently got a custom url validation regex in the project's Image model.

File: test/controllers/users_controller_test.rb (hunk 15 above)
Comment (bboe) [reply test/controllers/users_controller_test.rb]: The way you've done this is perfectly fine. Rails should use consistent casing, so I wouldn't worry about that. I might have actually combined the two `new` test cases as I consider the body containing the right form method and path to be part of the expected response.

File: test/controllers/users_controller_test.rb [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+  end
+
+  def test_create__fail
+    @user_params[:email] = 'invalid_email'
+    assert_no_difference('User.count') do
+      post users_path, params: { user: @user_params }
+    end
+    # Verify that we got an HTTP 400 response and that we're still on the Sign Up form
+    assert_response :bad_request
+    assert_template :new
```
Comment (bboe) [inline test/controllers/users_controller_test.rb]: Consider avoiding the use of `assert_template`. While it is utilized heavily as part of apm_bundle, it was removed from rails core due to considering implementation specifics rather than actual outcome. Instead consider what is important that is rendered by that template (the form again).

File: test/models/user_test.rb (hunk 7 above)
Comment (bboe) [reply test/models/user_test.rb]: > 4.5.3.1.1.  Local-part
> The maximum total length of a user name or other local-part is 64 octets.

https://tools.ietf.org/html/rfc5321#section-4.5.3.1

File: db/migrate/20171108182257_create_users.rb (hunk 1 above)
Comment (bboe) [reply db/migrate/20171108182257_create_users.rb]: :)

File: test/controllers/users_controller_test.rb (hunk 16 above)
Comment (bboe) [reply test/controllers/users_controller_test.rb]: Yes that is an instance variable. The approach you've taken is a great one. Another way to do that would be to write a factory method that uses valid default keyword arguments and returns the hash. Something like:

    def user_params(name: 'valid name', email: 'valid@email.com', password: 'password', password_confirmation: 'password')
        { name: name, email: email, password: password, password_confirmation: password }
    end

In the end they pretty much accomplish the same thing. What I personally prefer about the factory method is that I can overwrite individual fields in the same line.

File: test/controllers/users_controller_test.rb (hunk 17 above)
Comment (bboe) [reply test/controllers/users_controller_test.rb]: Fantastic. That's precisely what I was hoping would happen.

File: test/fixtures/users.yml (hunk 2 above)
Comment (bboe) [reply test/fixtures/users.yml]: It might be worth going through the exercise of trying to clean up your commits to get more practice with the interactive rebases. Git mastery is a skill few of our developers use to have, and through Ropes most everyone now has that mastery.

It is rare that we ever really _need_ to do such rebases, but I often ask my teammates to do so, in order to make PRs easier to review. I am a reviewer that likes to go commit-by-commit, which doesn't work well if lines are added in a commit and later removed by a subsequent commit. I also encourage others to review PRs in a similar way as opposed to looking at the entire diff.

File: test/models/user_test.rb [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+    @user[:email] = 'JON@Example.com'
+    @user.save!
+    assert_equal 'jon@example.com', @user.email
+  end
+
+  def test_user_name_required
+    @user[:name] = ''
+    assert_predicate @user, :invalid?
+    assert_equal 1, @user.errors.count
+    assert_equal "can't be blank", @user.errors.messages[:name].first
```
Comment (bboe) [inline test/models/user_test.rb]: Consider:

     assert_equal({name: ["can't be blank"]}, @user.errors.messages)

That provides complete precision in a single line.

Comment (jschneid) [reply test/models/user_test.rb]: What great syntax! Thanks for the suggestion!

Fixed throughout user_test.rb and image_test.rb in 7b99286.
